cpp/src/parquet/encryption/crypto_factory.cc

```diff
@@ -28,9 +28,38 @@
 
 namespace parquet::encryption {
 
+namespace {
+
+/// Holds a FileKeyUnwrapper and shared pointer to a KeyToolkit to allow
+/// keeping the KeyToolkit alive alongside the FileKeyUnwrapper
+class CryptoFactoryFileKeyRetriever : public DecryptionKeyRetriever {
+ public:
+  CryptoFactoryFileKeyRetriever(
+      std::shared_ptr<KeyToolkit> key_toolkit,
+      const KmsConnectionConfig& kms_connection_config, double cache_lifetime_seconds,
+      const std::string& file_path = "",
+      const std::shared_ptr<::arrow::fs::FileSystem>& file_system = NULLPTR)
+      : file_key_unwrapper_(key_toolkit.get(), kms_connection_config,
+                            cache_lifetime_seconds, file_path, file_system) {
+    key_toolkit_ = key_toolkit;
+  }
+
+  std::string GetKey(const std::string& key_metadata) override {
+    return file_key_unwrapper_.GetKey(key_metadata);
+  }
+
+ private:
+  FileKeyUnwrapper file_key_unwrapper_;
+  std::shared_ptr<KeyToolkit> key_toolkit_;
+};
+
+}  // namespace
+
+CryptoFactory::CryptoFactory() { key_toolkit_ = std::make_shared<KeyToolkit>(); }
```

Comment:
nit: move the assignment to initialization list, or simply initializing it in the header file to get rid of the explicit default constructor.